ddtrace/contrib/asyncpg/patch.py

```diff
@@ -0,0 +1,151 @@
+from typing import TYPE_CHECKING
+
+from ddtrace import Pin
+from ddtrace import config
+from ddtrace.vendor import wrapt
+
+from ...constants import SPAN_MEASURED_KEY
+from ...ext import SpanTypes
+from ...ext import db
+from ...ext import net
+from ...internal.logger import get_logger
+from ...internal.utils import get_argument_value
+from ..trace_utils import ext_service
+from ..trace_utils import unwrap
+from ..trace_utils import with_traced_module as with_traced_module_sync
+from ..trace_utils import wrap
+from ..trace_utils_async import with_traced_module
+
+
+if TYPE_CHECKING:
+    from types import ModuleType
+    from typing import Dict
+
+    import asyncpg
+
+config._add(
+    "asyncpg",
+    dict(
+        _default_service="postgres",
+    ),
+)
+
+
+log = get_logger(__name__)
+
+
+def _get_connection_tags(conn):
+    # type: (asyncpg.Connection) -> Dict[str, str]
+    addr = conn._addr
+    params = conn._params
+    host = port = ""
+    if isinstance(addr, tuple) and len(addr) == 2:
+        host, port = addr
+    return {
+        net.TARGET_HOST: host,
+        net.TARGET_PORT: port,
+        db.USER: params.user,
+        db.NAME: params.database,
+    }
+
+
+class _TracedConnection(wrapt.ObjectProxy):
+    def __init__(self, conn, pin):
+        super(_TracedConnection, self).__init__(conn)
+        pin.onto(self)
+
+    async def _traced_method(self, method, query, args, kwargs):
+        pin = Pin.get_from(self)
+        if not pin or not pin.enabled:
+            log.debug("No pin or pin disabled, skipping tracing for %s", method)
+            return await method(*args, **kwargs)
+
+        with pin.tracer.trace(
+            "postgres.query", resource=query, service=ext_service(pin, config.asyncpg), span_type=SpanTypes.SQL
+        ) as span:
+            span.set_tag(SPAN_MEASURED_KEY)
+            span.set_tags(_get_connection_tags(self))
+            return await method(*args, **kwargs)
+
+    async def execute(self, *args, **kwargs):
+        query = get_argument_value(args, kwargs, 0, "query")
+        return await self._traced_method(self.__wrapped__.execute, query, args, kwargs)
+
+    async def executemany(self, *args, **kwargs):
+        command = get_argument_value(args, kwargs, 0, "command")
+        return await self._traced_method(self.__wrapped__.executemany, command, args, kwargs)
+
+    async def fetch(self, *args, **kwargs):
+        query = get_argument_value(args, kwargs, 0, "query")
+        return await self._traced_method(self.__wrapped__.fetch, query, args, kwargs)
+
+    async def fetchval(self, *args, **kwargs):
+        query = get_argument_value(args, kwargs, 0, "query")
+        return await self._traced_method(self.__wrapped__.fetchval, query, args, kwargs)
+
+    async def fetchrow(self, *args, **kwargs):
+        query = get_argument_value(args, kwargs, 0, "query")
+        return await self._traced_method(self.__wrapped__.fetchrow, query, args, kwargs)
```

Comment:
These should be disabled by default just like dbapi.

these are going to be _very_ noisy.

Reply:
wait, right? but these take a query?

Reply:
yeah it's weird because they take a query, these fetch methods are slightly different than cursor fetch methods